bigtable/google/cloud/bigtable/row_set.py

```diff
@@ -0,0 +1,175 @@
+            message.rows.row_ranges.add(**r_kwrags)
+
+
+class RowRange(object):
+    """ Convenience wrapper of google.bigtable.v2.RowRange
+
+    :type start_key: string utf-8
+    :param start_key: start key of the row range
+
+    :type end_key: string utf-8
+    :param end_key: end key of the row range
+
+    :type start_inclusive: bool
+    :param start_inclusive: (Optional) Whether the ``start_key`` should be
+                  considered inclusive. The default is True (inclusive).
+
+    :type end_inclusive: bool
+    :param end_inclusive: (Optional) Whether the ``end_key`` should be
+                  considered inclusive. The default is False (exclusive).
+    """
+
+    def __init__(self, start_key, end_key,
+                 start_inclusive=True,
+                 end_inclusive=False):
+        self.start_key = start_key
+        self.start_inclusive = start_inclusive
+        self.end_key = end_key
+        self.end_inclusive = end_inclusive
+
+    def get_range_kwargs(self):
+        """ Convert row range object to dict which can be passed to
+        google.bigtable.v2.RowRange add method.
+        """
+        range_kwargs = {}
+        start_key_key = 'start_key_open'
+        if self.start_inclusive:
+            start_key_key = 'start_key_closed'
+        range_kwargs[start_key_key] = _to_bytes(self.start_key)
+        end_key_key = 'end_key_open'
+        if self.end_inclusive:
+            end_key_key = 'end_key_closed'
+        range_kwargs[end_key_key] = _to_bytes(self.end_key)
+        return range_kwargs
+
+
+class ReadRowsRequestManager(object):
+    """ Update the ReadRowsRequest message in case of failures by
+        filtering the already read keys.
+
+    :type message: class:`data_messages_v2_pb2.ReadRowsRequest`
+    :param message: Original ReadRowsRequest containing all of the parameters
+                    of API call
+
+    :type last_scanned_key: bytes
+    :param last_scanned_key: last successfully scanned key
+
+    :type rows_read_so_far: int
+    :param rows_read_so_far: total no of rows successfully read so far.
+                            this will be used for updating rows_limit
+
+    """
+
+    def __init__(self, message, last_scanned_key, rows_read_so_far):
+        self.message = message
+        self.last_scanned_key = last_scanned_key
+        self.rows_read_so_far = rows_read_so_far
+        self.new_message = None
+
+    def build_updated_request(self):
+        """ Updates the given message request as per last scanned key
+        """
+        r_kwargs = {'table_name': self.message.table_name}
+        r_kwargs['filter'] = self.message.filter
+
+        if self.message.rows_limit != 0:
+            r_kwargs['rows_limit'] = max(1, self.message.rows_limit -
+                                         self.rows_read_so_far)
+
+        self.new_message = data_messages_v2_pb2.ReadRowsRequest(**r_kwargs)
+        self._filter_rows_keys()
+        self._filter_row_ranges()
```

Comment:
Rather than scribbling on an already-created message, couldn't we just generate the `row_keys` and `row_ranges` ourselves, and then populate the message with the corresponding `RowSet`?  E.g.:

```python
    row_keys = [
        row_key for row_key in self.message.rows.row_keys
         if row_key > self.last_scanned_key]
    row_ranges = self._filter_row_ranges()
    r_kwargs['row_set'] = data_v2_pb2.RowSet(
        row_keys=row_keys, row_ranges=row_ranges)
    return data_messages_v2_pb2.ReadRowsRequest(**r_kwargs)

Reply:
Changes made as per your suggestion. Thanks.